src/box/xlog.c

```diff
@@ -1357,6 +1360,13 @@ xlog_write_row(struct xlog *log, const struct xrow_header *packet)
 	    xlog_tx_write(log) < 0)
 		return -1;
 
+	struct errinj *inj = errinj(ERRINJ_WAL_WRITE_TIMEOUT, ERRINJ_DOUBLE);
+	if (inj != NULL && inj->dparam > 0) {
+		thread_sleep(inj->dparam);
+		if (loop())
+			ev_now_update(loop());
```

Comment:
Please don't call `ev_now_update` on error injection. Move the main call so that it's called in any case.

Reply:
Moved back to yield based on iteration counter as discussed f2f.

<details>
<summary>diff</summary>

```diff
diff --git a/src/box/memtx_engine.cc b/src/box/memtx_engine.cc
index abb0b3d67..e13c6b906 100644
--- a/src/box/memtx_engine.cc
+++ b/src/box/memtx_engine.cc
@@ -68,9 +68,6 @@
 /* sync snapshot every 16MB */
 #define SNAP_SYNC_INTERVAL	(1 << 24)
 
-/* How frequently checkpoint thread yields. */
-#define CHECKPOINT_YIELD_PERIOD 0.1f
-
 enum {
 	OBJSIZE_MIN = 16,
 	SLAB_SIZE = 16 * 1024 * 1024,
@@ -764,6 +761,8 @@ checkpoint_write_tuple(struct xlog *l, uint32_t space_id, uint32_t group_id,
 	row.body[0].iov_len = sizeof(body);
 	row.body[1].iov_base = (char *)data;
 	row.body[1].iov_len = size;
+	ERROR_INJECT_DOUBLE(ERRINJ_SNAP_WRITE_TIMEOUT, inj->dparam > 0,
+			    thread_sleep(inj->dparam));
 	return checkpoint_write_row(l, &row);
 }
 
@@ -985,6 +984,12 @@ checkpoint_write_invalid_system_row(struct xlog *l)
 static int
 checkpoint_f(va_list ap)
 {
+#ifdef NDEBUG
+	enum { YIELD_LOOPS = 1000 };
+#else
+	enum { YIELD_LOOPS = 10 };
+#endif
+
 	int rc = 0;
 	struct checkpoint *ckpt = va_arg(ap, struct checkpoint *);
 
@@ -1040,7 +1045,7 @@ checkpoint_f(va_list ap)
 			rc = -1;
 			break;
 		}
-		double last_yield_time = ev_monotonic_now(loop());
+		unsigned int loops;
 		while (true) {
 			RegionGuard region_guard(&fiber()->gc);
 			struct read_view_tuple result;
@@ -1057,11 +1062,8 @@ checkpoint_f(va_list ap)
 			if (rc != 0)
 				break;
 			/* Yield to make thread cancellable. */
-			if (ev_monotonic_now(loop()) - last_yield_time >
-			    CHECKPOINT_YIELD_PERIOD) {
-				last_yield_time = ev_monotonic_now(loop());
+			if (++loops % YIELD_LOOPS == 0)
 				fiber_sleep(0);
-			}
 			if (fiber_is_cancelled()) {
 				diag_set(FiberIsCancelled);
 				rc = -1;
diff --git a/src/box/xlog.c b/src/box/xlog.c
index 9b768ee57..5b7c550eb 100644
--- a/src/box/xlog.c
+++ b/src/box/xlog.c
@@ -1230,9 +1230,6 @@ xlog_tx_write(struct xlog *log)
 	} else {
 		written = xlog_tx_write_plain(log);
 	}
-	/* In case we want to yield depending on how much time pass. */
-	if (loop())
-		ev_now_update(loop());
 	ERROR_INJECT(ERRINJ_WAL_WRITE, {
 		diag_set(ClientError, ER_INJECTION, "xlog write injection");
 		written = -1;
@@ -1360,13 +1357,6 @@ xlog_write_row(struct xlog *log, const struct xrow_header *packet)
 	    xlog_tx_write(log) < 0)
 		return -1;
 
-	struct errinj *inj = errinj(ERRINJ_WAL_WRITE_TIMEOUT, ERRINJ_DOUBLE);
-	if (inj != NULL && inj->dparam > 0) {
-		thread_sleep(inj->dparam);
-		if (loop())
-			ev_now_update(loop());
-	}
-
 	return row_size;
 }
 
diff --git a/src/lib/core/errinj.h b/src/lib/core/errinj.h
index e35857030..f510564b4 100644
--- a/src/lib/core/errinj.h
+++ b/src/lib/core/errinj.h
@@ -133,6 +133,7 @@ struct errinj {
 	_(ERRINJ_SNAP_WRITE_CORRUPTED_INSERT_ROW, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_SNAP_WRITE_INVALID_SYSTEM_ROW, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_SNAP_WRITE_MISSING_SPACE_ROW, ERRINJ_BOOL, {.bparam = false}) \
+	_(ERRINJ_SNAP_WRITE_TIMEOUT, ERRINJ_DOUBLE, {.dparam = 0}) \
 	_(ERRINJ_SNAP_WRITE_UNKNOWN_ROW_TYPE, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_SPACE_UPGRADE_DELAY, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_SWIM_FD_ONLY, ERRINJ_BOOL, {.bparam = false}) \
@@ -185,7 +186,6 @@ struct errinj {
 	_(ERRINJ_WAL_WRITE_DISK, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_WAL_WRITE_EOF, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_WAL_WRITE_PARTIAL, ERRINJ_INT, {.iparam = -1}) \
-	_(ERRINJ_WAL_WRITE_TIMEOUT, ERRINJ_DOUBLE, {.dparam = 0}) \
 	_(ERRINJ_XLOG_GARBAGE, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_XLOG_META, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_XLOG_READ, ERRINJ_INT, {.iparam = -1}) \
diff --git a/test/app-luatest/shutdown_test.lua b/test/app-luatest/shutdown_test.lua
index f7b7c1150..b2c02cfab 100644
--- a/test/app-luatest/shutdown_test.lua
+++ b/test/app-luatest/shutdown_test.lua
@@ -139,10 +139,12 @@ g_snap.test_shutdown_during_memtx_snapshot = function(cg)
         local fiber = require('fiber')
         box.schema.create_space('test')
         box.space.test:create_index('pk')
+        box.begin()
         for i=1,10000 do
             box.space.test:insert{i}
         end
-        box.error.injection.set('ERRINJ_WAL_WRITE_TIMEOUT', 0.1)
+        box.commit()
+        box.error.injection.set('ERRINJ_SNAP_WRITE_TIMEOUT', 0.01)
         fiber.create(function()
             box.snapshot()
         end)
diff --git a/test/box/errinj.result b/test/box/errinj.result
index e1950292c..af5dcc6ce 100644
--- a/test/box/errinj.result
+++ b/test/box/errinj.result
@@ -106,6 +106,7 @@ evals
   - ERRINJ_SNAP_WRITE_DELAY: false
   - ERRINJ_SNAP_WRITE_INVALID_SYSTEM_ROW: false
   - ERRINJ_SNAP_WRITE_MISSING_SPACE_ROW: false
+  - ERRINJ_SNAP_WRITE_TIMEOUT: 0
   - ERRINJ_SNAP_WRITE_UNKNOWN_ROW_TYPE: false
   - ERRINJ_SPACE_UPGRADE_DELAY: false
   - ERRINJ_SWIM_FD_ONLY: false
@@ -158,7 +159,6 @@ evals
   - ERRINJ_WAL_WRITE_DISK: false
   - ERRINJ_WAL_WRITE_EOF: false
   - ERRINJ_WAL_WRITE_PARTIAL: -1
-  - ERRINJ_WAL_WRITE_TIMEOUT: 0
   - ERRINJ_XLOG_GARBAGE: false
   - ERRINJ_XLOG_META: false
   - ERRINJ_XLOG_READ: -1

```

</details>